File: app/src/main/java/net/xzos/UpgradeAll/ui/viewmodels/adapters/CloudHubItemAdapter.java
package net.xzos.UpgradeAll.ui.viewmodels.adapters;

import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.PopupMenu;
import androidx.recyclerview.widget.RecyclerView;

import net.xzos.UpgradeAll.R;
import net.xzos.UpgradeAll.gson.HubConfig;
import net.xzos.UpgradeAll.server.hub.CloudHub;
import net.xzos.UpgradeAll.server.hub.HubManager;
import net.xzos.UpgradeAll.ui.viewmodels.ItemCardView;

import java.util.List;


public class CloudHubItemAdapter extends RecyclerView.Adapter<CardViewRecyclerViewHolder> {

    private static final String TAG = "CloudHubItemAdapter";

    private List<ItemCardView> mItemCardViewList;

    private CloudHub mCloudHub;

    public CloudHubItemAdapter(List<ItemCardView> updateList, CloudHub cloudHub) {
        mItemCardViewList = updateList;
        mCloudHub = cloudHub;
    }

    @NonNull
    @Override
    public CardViewRecyclerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final CardViewRecyclerViewHolder holder = new CardViewRecyclerViewHolder(
                LayoutInflater.from(parent.getContext()).inflate(R.layout.cardview_item, parent, false));
        // 长按菜单
        holder.itemCardView.setOnLongClickListener(v -> {
            final int position = holder.getAdapterPosition();
            final ItemCardView itemCardView = mItemCardViewList.get(position);
            PopupMenu popupMenu = new PopupMenu(holder.itemCardView.getContext(), v);
            MenuInflater menuInflater = popupMenu.getMenuInflater();
            menuInflater.inflate(R.menu.menu_long_click_cardview_item_cloud_hub, popupMenu.getMenu());
            popupMenu.show();
            //设置item的点击事件
            popupMenu.setOnMenuItemClickListener(item -> {
                switch (item.getItemId()) {
                    // 下载
                    case R.id.download:
                        Toast.makeText(holder.itemCardView.getContext(), "开始下载", Toast.LENGTH_LONG).show();
                        // 下载数据
                        new Thread(() -> {
                            HubConfig cloudHubConfigGson = mCloudHub.getHubConfig(itemCardView.getExtraData().getConfigFileName());
                            boolean addHubSuccess;
                            if (cloudHubConfigGson != null) {
                                String cloudHubConfigJS = mCloudHub.getHubConfigJS(cloudHubConfigGson.getWebCrawler().getFilePath());
                                if (cloudHubConfigJS != null)
                                    addHubSuccess = HubManager.addHubDatabase(0, cloudHubConfigGson, cloudHubConfigJS);
                                else {
                                    addHubSuccess = false;
                                    new Handler(Looper.getMainLooper()).post(() -> {
                                        Toast.makeText(holder.itemCardView.getContext(), "获取 JS 代码失败", Toast.LENGTH_LONG).show();
                                    });
                                }
                            } else {
                                addHubSuccess = false;
                                new Handler(Looper.getMainLooper()).post(() -> {
                                    Toast.makeText(holder.itemCardView.getContext(), "数据下载失败", Toast.LENGTH_LONG).show();
                                });
                            }
                            new Handler(Looper.getMainLooper()).post(() -> {
                                if (addHubSuccess) {
                                    Toast.makeText(holder.itemCardView.getContext(), "数据添加成功", Toast.LENGTH_LONG).show();
                                } else
                                    Toast.makeText(holder.itemCardView.getContext(), "什么？数据库添加失败！", Toast.LENGTH_LONG).show();
                            });
                        }).start();
                        // 添加数据库
                        break;
                }
                return true;
            });
            return true;
        });
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull CardViewRecyclerViewHolder holder, int position) {
        ItemCardView itemCardView = mItemCardViewList.get(position);
        // 底栏设置
        if (itemCardView.getExtraData().isEmpty()) {
            holder.itemCardView.setVisibility(View.GONE);
            holder.endTextView.setVisibility(View.VISIBLE);
        } else {
            holder.itemCardView.setVisibility(View.VISIBLE);
            holder.endTextView.setVisibility(View.GONE);
            holder.name.setText(itemCardView.getName());
            holder.api.setText(itemCardView.getApi());
            holder.descTextView.setText(itemCardView.getDesc());
            holder.descTextView.setEnabled(false);
        }
    }

    @Override
    public int getItemCount() {
        return mItemCardViewList.size();
    }
}

File: app/src/main/java/net/xzos/upgradeall/ui/viewmodels/adapters/CloudHubItemAdapter.kt
package net.xzos.upgradeall.ui.viewmodels.adapters

import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.widget.PopupMenu
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import net.xzos.upgradeall.R
import net.xzos.upgradeall.core.data_manager.CloudConfigGetter
import net.xzos.upgradeall.core.data_manager.HubDatabaseManager
import net.xzos.upgradeall.ui.viewmodels.view.CloudConfigListItemView
import net.xzos.upgradeall.ui.viewmodels.view.holder.CardViewRecyclerViewHolder
import net.xzos.upgradeall.utils.IconPalette


class CloudHubItemAdapter(override var mItemCardViewList: List<CloudConfigListItemView>)
    : CloudItemAdapter(mItemCardViewList) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CardViewRecyclerViewHolder {
        val holder = super.onCreateViewHolder(parent, viewType)
        // 长按菜单
        holder.itemCardView.setOnLongClickListener { v ->
            val position = holder.adapterPosition
            val itemCardView = mItemCardViewList[position]
            val popupMenu = PopupMenu(holder.itemCardView.context, v)
            val menuInflater = popupMenu.menuInflater
            menuInflater.inflate(R.menu.menu_long_click_cardview_item_cloud_hub, popupMenu.menu)
            popupMenu.show()
            //设置item的点击事件
            popupMenu.setOnMenuItemClickListener { item ->
                if (item.itemId == R.id.download) {
                    setDownloadStatus(holder, true)
                    GlobalScope.launch {
                        val addHubStatus = CloudConfigGetter.downloadCloudHubConfig(itemCardView.uuid)  // 下载数据
                        withContext(Dispatchers.Main) {
                            setDownloadStatus(holder, false)
                            if (addHubStatus == 3) {
                                holder.versionCheckButton.visibility = View.VISIBLE
                            }
                        }
                    }
                }
                return@setOnMenuItemClickListener true
            }
            return@setOnLongClickListener true
        }
        return holder
    }

    override fun onBindViewHolder(holder: CardViewRecyclerViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)
        val itemCardView = mItemCardViewList[position]
        if (itemCardView.uuid != null) {
            // 加载仓库信息
            holder.nameTextView.text = itemCardView.name
            itemCardView.type?.let {
                holder.typeTextView.setText(it)
            }
            holder.hubNameTextView.visibility = View.GONE
            checkHubConfigLocalStatus(holder, itemCardView.uuid)
            GlobalScope.launch { loadCloudHubIcon(holder.appIconImageView, itemCardView.uuid) }
        }
    }

    private fun loadCloudHubIcon(iconImageView: ImageView, hubUuid: String?) {
        if (hubUuid != null) {
            val cloudHubConfigGson = CloudConfigGetter.getHubCloudConfig(hubUuid)
            val hubIconUrl = cloudHubConfigGson?.info?.hubIconUrl
            IconPalette.loadHubIconView(iconImageView, hubIconUrl)
        }
    }

    private fun checkHubConfigLocalStatus(holder: CardViewRecyclerViewHolder, hubUuid: String?) {
        val versionCheckButton = holder.versionCheckButton
        if (HubDatabaseManager.exists(hubUuid)) {
            versionCheckButton.visibility = View.VISIBLE
            setDownloadStatus(holder, true)
            val hubConfigGson = CloudConfigGetter.getHubCloudConfig(hubUuid)
            HubDatabaseManager.getDatabase(uuid = hubUuid)?.hubConfig?.let {
                val localHubVersion = it.info.configVersion
                val cloudHubVersion = hubConfigGson?.info?.configVersion
                if (cloudHubVersion != null && cloudHubVersion > localHubVersion)
                    versionCheckButton.setImageResource(R.drawable.ic_check_needupdate)
                setDownloadStatus(holder, false)
            }
        } else
            versionCheckButton.visibility = View.GONE
    }
}
